FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+    we expose their gain configuration?
+
+## Using SearchIndexingBufferedSender
+
+- `public class SearchIndexingBufferedSender<T>`
+    - Add Indexing Actions
+        - `void DeleteDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task DeleteDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void IndexDocuments(IndexDocumentsBatch<T>, CancellationToken)`
+        - `Task IndexDocumentsAsync(IndexDocumentsBatch<T>, CancellationToken)`
+
+            Note these are using a batch which we already have conveniences for
+            building up.
+
+        - `void MergeDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task MergeDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void MergeOrUploadDocuments(IEnumerable, CancellationToken)`
+        - `Task MergeOrUploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void UploadDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task UploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+
+        We're not returning any data from these operations.  They may submit
+        something immediately or we might be waiting for a flush.
+
+        The provided cancellation token will be used for any requests sent, any
+        notifications raised, etc.  If a flush happens when the timer elapses,
+        then we'll use the `AutoFlushIntervalCancellationToken` instead.  A 
+        customer could reuse a token or link them together for fine grained
+        control.
+
+        We're also not including the `IndexDocumentsOptions` parameter because
+        that currently only serves to configure exception behavior which we're
+        doing exclusively through events.
+
+        - Delete by key
+
+            Add customization for "delete with just a key" scenarios for dynamic
+            `SearchDocument`.  We're using extension methods in .NET but other
+            languages might need to figure out a different approach.
+
+            - `public static class SearchExtensions`
+                - `static void DeleteDocuments(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
+                - `static Task DeleteDocumentsAsync(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
+
+    - `void Flush(CancellationToken)` and `Task FlushAsync(CancellationToken)`
+        will manually flush any pending operations.  This could result in
+        multiple requests.  If a request is already in progress, whatever's left
+        will still be flushed on returning.
+
+    - Notifications will be raised as events.
+        - `event Action<IndexDocumentsAction<T>, CancellationToken> ActionAdded;`
+        - `event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionAddedAsync;`
+        - `event Action<IndexDocumentsAction<T>, CancellationToken> ActionSent;`
+        - `event Func<IndexDocumentsAction<T>, CancellationToken, Task> ActionSentAsync;`
```

Comment:
Should we have this send the entire batch or send it once per document?  Some of the other events make checkpointing easier with once per document so we'll mirror that here even though there's not a huge difference once way or the other.